conda_forge_tick/git_utils.py

```diff
@@ -88,77 +97,533 @@ def github_client() -> github.Github:
+            return
+
+        repo = self.github3_client.repository(owner, repo_name)
+        if repo is None:
+            raise RepositoryNotFoundError(
+                f"Repository {owner}/{repo_name} does not exist."
+            )
+
+        logger.debug(f"Forking {owner}/{repo_name}.")
+        repo.create_fork()
+
+        # Sleep to make sure the fork is created before we go after it
+        time.sleep(5)
+
+    def _sync_default_branch(self, upstream_owner: str, repo_name: str):
+        fork_owner = self.user
+
+        upstream_repo = self.pygithub_client.get_repo(f"{upstream_owner}/{repo_name}")
+        fork = self.pygithub_client.get_repo(f"{fork_owner}/{repo_name}")
+
+        if upstream_repo.default_branch == fork.default_branch:
+            return
+
+        logger.info(
+            f"Syncing default branch of {fork_owner}/{repo_name} with {upstream_owner}/{repo_name}..."
+        )
+
+        fork.rename_branch(fork.default_branch, upstream_repo.default_branch)
+
+        # Sleep to wait for branch name change
+        time.sleep(5)
+
+    @cached_property
+    def user(self) -> str:
+        return self.pygithub_client.get_user().login
+
+    def get_api_requests_left(self) -> int | None:
+        try:
+            limit_info = self.github3_client.rate_limit()
+        except github3.exceptions.GitHubException as e:
+            logger.warning("GitHub API error while fetching rate limit.", exc_info=e)
+            return None
+
+        try:
+            core_resource = limit_info["resources"]["core"]
+            remaining_limit = core_resource["remaining"]
+        except KeyError as e:
+            logger.warning("GitHub API error while parsing rate limit.", exc_info=e)
+            return None
+
+        if remaining_limit != 0:
+            return remaining_limit
+
+        # try to log when the limit will be reset
+        try:
+            reset_timestamp = core_resource["reset"]
+        except KeyError as e:
+            logger.warning(
+                "GitHub API error while fetching rate limit reset time.",
+                exc_info=e,
+            )
+            return remaining_limit
+
+        logger.info(
+            "GitHub API limit reached, will reset at "
+            f"{datetime.utcfromtimestamp(reset_timestamp).strftime('%Y-%m-%dT%H:%M:%SZ')}"
+        )
+
+        return remaining_limit
+
+
+class DryRunBackend(GitPlatformBackend):
+    """
+    A git backend that doesn't modify anything and only relies on public APIs that do not require authentication.
+    Useful for local testing with dry-run.
+    """
+
+    _USER = "auto-tick-bot-dry-run"
+
+    def __init__(self):
+        super().__init__(GitCli())
+        self._repos: set[str] = set()
+
+    def get_api_requests_left(self) -> Bound:
+        return Bound.INFINITY
+
+    def does_repository_exist(self, owner: str, repo_name: str) -> bool:
+        if owner == self._USER:
+            return repo_name in self._repos
```

Comment:
I think that's the point though. This method tests if the fork exists so will return False if it does not.

Reply:
Yes, exactly. Note that this does not prevent us from calling `does_repository_exist` with `regro-cf-autotick-bot` (the actual bot user) as repository owner since `_USER` is a special placeholder value.

All forks are created by the virtual `_USER` user, simulating GitHub behaviour.